rosette/h/MI.h

```diff
@@ -70,115 +58,88 @@ class ProductType : public Actor {
 
-inline int ProductType::numberOfElements() {
-    return definite()->numberOfElements();
-}
+    pOb elemR(int i) {
+        if (i < definite()->numberOfElements()) {
+            return elem(i);
+        } else {
+            return star();
+        }
+    }
+};
 
-inline pOb ProductType::elem(int i) { return definite()->elem(i); }
 
-inline pOb ProductType::elemR(int i) {
-    if (i < definite()->numberOfElements())
-        return elem(i);
-    else
-        return star();
-}
+static const int SUM_TYPE_TYPES_SLOT = 0;
+static const int BUILTIN_SumType_SLOTS = 1;
 
 
 class SumType : public Actor {
     STD_DECLS(SumType);
 
-   protected:
+    protected:
     SumType(pExt);
 
-   public:
+    public:
     static SumType* create(Tuple*);
-    Tuple* types();
-    pOb elem(int);
-    int numberOfElements();
     virtual bool compositeCoversp(pOb);
     virtual bool isCoveredByp(pOb);
     pOb dominator();
     virtual pOb typeLub(pOb);
     pOb normalize();
-};
-
 
-static const int SUM_TYPE_TYPES_SLOT = 0;
+    Tuple* types() {
+        return (Tuple*)extension->slot(SUM_TYPE_TYPES_SLOT);
+    }
 
-static const int BUILTIN_SumType_SLOTS = 1;
-
-inline Tuple* SumType::types() {
-    return (Tuple*)extension->slot(SUM_TYPE_TYPES_SLOT);
-}
-
-inline pOb SumType::elem(int i) { return types()->elem(i); }
+    pOb elem(int i) { return types()->elem(i); }
+    int numberOfElements() { return types()->numberOfElements(); }
+};
 
-inline int SumType::numberOfElements() { return types()->numberOfElements(); }
 
+static const int MM_PROC_LIST_SLOT = 0;
+static const int BUILTIN_MultiMethod_SLOTS = 1;
 
 class MultiMethod : public Actor {
     STD_DECLS(MultiMethod);
 
-   protected:
+    protected:
     MultiMethod(pExt);
 
-   public:
+    public:
     static MultiMethod* create();
-    Tuple* procList();
-    pOb elem(int);
-    int numberOfElements();
     pOb matchAndDispatch(pCtxt);
-};
-
-
-static const int MM_PROC_LIST_SLOT = 0;
 
-static const int BUILTIN_MultiMethod_SLOTS = 1;
-
-inline Tuple* MultiMethod::procList() {
-    return (Tuple*)extension->slot(MM_PROC_LIST_SLOT);
-}
-
-inline pOb MultiMethod::elem(int i) { return procList()->elem(i); }
+    Tuple* procList() {
+        return (Tuple*)extension->slot(MM_PROC_LIST_SLOT);
+    }
 
-inline int MultiMethod::numberOfElements() {
-    return procList()->numberOfElements();
-}
+    pOb elem(int i) { return procList()->elem(i); }
 
-inline bool typeLessEq(pOb x, pOb y) { return BASE(y)->coversp(x); }
+    int numberOfElements() {
+        return procList()->numberOfElements();
+    }
 
-inline bool typeGreaterEq(pOb x, pOb y) { return BASE(x)->coversp(y); }
-
-inline bool typeEq(pOb x, pOb y) {
-    return (BASE(x)->coversp(y) && BASE(y)->coversp(x));
-}
-
-inline bool typeLess(pOb x, pOb y) {
-    return (typeLessEq(x, y) && !typeEq(x, y));
-}
+};
 
-inline bool typeGreater(pOb x, pOb y) {
-    return (typeGreaterEq(x, y) && !typeEq(x, y));
-}
+bool typeLessEq(pOb x, pOb y);
```

Comment:
Same as above. You've now made these non-inline.

Reply:
Yeah, they don't compile anyway because the header situation is a mess and you end up with multiple definitions. This is what #pragma interface was overriding, but it's gross and makes the code harder to deal with.

Reply:
You don't need #pragma interface to define an inline function in a header. That's standard C++.